### PR TITLE
lxd: Add VM image and server support for QEMU VMs

### LINKED ISSUE #114194: LXD Virtual-Machine (qemu) mode is not working
**Describe the bug**
Since lxd 4.0 lxd is able to run qemu virtual machines besides just containers
https://discuss.linuxcontainers.org/t/running-virtual-machines-with-lxd-4-0/7519/1

There seem to be hardcoded locations to some /usr/share foloders that are missing, etc

Additionally not sure about qemu inclusion into lxd

A variant of lxd could be built with a flag "useQemu ? false" in the package and a configuration option to toggle it could be added

**To Reproduce**
1. `lxc init win10 --empty --vm -c security.secureboot=false -c limits.cpu=4 -c limits.memory=4GB`
2. `lxc start win10`

**Expected behavior**
It tries to boot

**Screenshots**
```
# lxc start win10
Error: lstat /usr/share: no such file or directory
Try `lxc info --show-log win10` for more info
```

**Additional context**
/usr/share hardcoded stuff

**Notify maintainers**

@fpletz @wucke13 

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.97, NixOS, 21.05pre-git (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20201205_a5d85d0`
 - channels(root): `"nixos-21.03pre269343.758b29b5a28, nixpkgs, solaros"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### LINKED ISSUE #140431: lxd: Unable to create a virtual machine
### Describe the bug
Currently it seems to be unable to create a virtual machine with lxd.

```
asbachb@nixos-t14s  /etc/nixos/modules  lxc launch images:ubuntu/hirsute ubuntu-vm --vm

Creating ubuntu-vm
Error: Failed instance creation: Failed creating instance record: Instance type "virtual-machine" is not supported on this server
```

### Steps To Reproduce
1. Enable `lxd`:
```nix
  virtualisation = {
    lxd = {
      enable = true;
      recommendedSysctlSettings = true;
    };
  };
```
2. Ensure your user is in `lxd` group
3. Try to create a virtual machine: `lxc launch images:ubuntu/hirsute ubuntu-vm --vm`

### Additional Context
```console
asbachb@nixos-t14s  /etc/nixos/modules  lxc launch images:ubuntu/hirsute ubuntu-vm --vm --debug

DBUG[10-03|21:15:56] Connecting to a local LXD over a Unix socket
DBUG[10-03|21:15:56] Sending request to LXD                   method=GET url=http://unix.socket/1.0 etag=
DBUG[10-03|21:15:56] Got response struct from LXD
DBUG[10-03|21:15:56]
	{
		"config": {},
		"api_extensions": [
			"storage_zfs_remove_snapshots",
			"container_host_shutdown_timeout",
			"container_stop_priority",
			"container_syscall_filtering",
			"auth_pki",
			"container_last_used_at",
			"etag",
			"patch",
			"usb_devices",
			"https_allowed_credentials",
			"image_compression_algorithm",
			"directory_manipulation",
			"container_cpu_time",
			"storage_zfs_use_refquota",
			"storage_lvm_mount_options",
			"network",
			"profile_usedby",
			"container_push",
			"container_exec_recording",
			"certificate_update",
			"container_exec_signal_handling",
			"gpu_devices",
			"container_image_properties",
			"migration_progress",
			"id_map",
			"network_firewall_filtering",
			"network_routes",
			"storage",
			"file_delete",
			"file_append",
			"network_dhcp_expiry",
			"storage_lvm_vg_rename",
			"storage_lvm_thinpool_rename",
			"network_vlan",
			"image_create_aliases",
			"container_stateless_copy",
			"container_only_migration",
			"storage_zfs_clone_copy",
			"unix_device_rename",
			"storage_lvm_use_thinpool",
			"storage_rsync_bwlimit",
			"network_vxlan_interface",
			"storage_btrfs_mount_options",
			"entity_description",
			"image_force_refresh",
			"storage_lvm_lv_resizing",
			"id_map_base",
			"file_symlinks",
			"container_push_target",
			"network_vlan_physical",
			"storage_images_delete",
			"container_edit_metadata",
			"container_snapshot_stateful_migration",
			"storage_driver_ceph",
			"storage_ceph_user_name",
			"resource_limits",
			"storage_volatile_initial_source",
			"storage_ceph_force_osd_reuse",
			"storage_block_filesystem_btrfs",
			"resources",
			"kernel_limits",
			"storage_api_volume_rename",
			"macaroon_authentication",
			"network_sriov",
			"console",
			"restrict_devlxd",
			"migration_pre_copy",
			"infiniband",
			"maas_network",
			"devlxd_events",
			"proxy",
			"network_dhcp_gateway",
			"file_get_symlink",
			"network_leases",
			"unix_device_hotplug",
			"storage_api_local_volume_handling",
			"operation_description",
			"clustering",
			"event_lifecycle",
			"storage_api_remote_volume_handling",
			"nvidia_runtime",
			"container_mount_propagation",
			"container_backup",
			"devlxd_images",
			"container_local_cross_pool_handling",
			"proxy_unix",
			"proxy_udp",
			"clustering_join",
			"proxy_tcp_udp_multi_port_handling",
			"network_state",
			"proxy_unix_dac_properties",
			"container_protection_delete",
			"unix_priv_drop",
			"pprof_http",
			"proxy_haproxy_protocol",
			"network_hwaddr",
			"proxy_nat",
			"network_nat_order",
			"container_full",
			"candid_authentication",
			"backup_compression",
			"candid_config",
			"nvidia_runtime_config",
			"storage_api_volume_snapshots",
			"storage_unmapped",
			"projects",
			"candid_config_key",
			"network_vxlan_ttl",
			"container_incremental_copy",
			"usb_optional_vendorid",
			"snapshot_scheduling",
			"snapshot_schedule_aliases",
			"container_copy_project",
			"clustering_server_address",
			"clustering_image_replication",
			"container_protection_shift",
			"snapshot_expiry",
			"container_backup_override_pool",
			"snapshot_expiry_creation",
			"network_leases_location",
			"resources_cpu_socket",
			"resources_gpu",
			"resources_numa",
			"kernel_features",
			"id_map_current",
			"event_location",
			"storage_api_remote_volume_snapshots",
			"network_nat_address",
			"container_nic_routes",
			"rbac",
			"cluster_internal_copy",
			"seccomp_notify",
			"lxc_features",
			"container_nic_ipvlan",
			"network_vlan_sriov",
			"storage_cephfs",
			"container_nic_ipfilter",
			"resources_v2",
			"container_exec_user_group_cwd",
			"container_syscall_intercept",
			"container_disk_shift",
			"storage_shifted",
			"resources_infiniband",
			"daemon_storage",
			"instances",
			"image_types",
			"resources_disk_sata",
			"clustering_roles",
			"images_expiry",
			"resources_network_firmware",
			"backup_compression_algorithm",
			"ceph_data_pool_name",
			"container_syscall_intercept_mount",
			"compression_squashfs",
			"container_raw_mount",
			"container_nic_routed",
			"container_syscall_intercept_mount_fuse",
			"container_disk_ceph",
			"virtual-machines",
			"image_profiles",
			"clustering_architecture",
			"resources_disk_id",
			"storage_lvm_stripes",
			"vm_boot_priority",
			"unix_hotplug_devices",
			"api_filtering",
			"instance_nic_network",
			"clustering_sizing",
			"firewall_driver",
			"projects_limits",
			"container_syscall_intercept_hugetlbfs",
			"limits_hugepages",
			"container_nic_routed_gateway",
			"projects_restrictions",
			"custom_volume_snapshot_expiry",
			"volume_snapshot_scheduling",
			"trust_ca_certificates",
			"snapshot_disk_usage",
			"clustering_edit_roles",
			"container_nic_routed_host_address",
			"container_nic_ipvlan_gateway",
			"resources_usb_pci",
			"resources_cpu_threads_numa",
			"resources_cpu_core_die",
			"api_os",
			"container_nic_routed_host_table",
			"container_nic_ipvlan_host_table",
			"container_nic_ipvlan_mode",
			"resources_system",
			"images_push_relay",
			"network_dns_search",
			"container_nic_routed_limits",
			"instance_nic_bridged_vlan",
			"network_state_bond_bridge",
			"usedby_consistency",
			"custom_block_volumes",
			"clustering_failure_domains",
			"resources_gpu_mdev",
			"console_vga_type",
			"projects_limits_disk",
			"network_type_macvlan",
			"network_type_sriov",
			"container_syscall_intercept_bpf_devices",
			"network_type_ovn",
			"projects_networks",
			"projects_networks_restricted_uplinks",
			"custom_volume_backup",
			"backup_override_name",
			"storage_rsync_compression",
			"network_type_physical",
			"network_ovn_external_subnets",
			"network_ovn_nat",
			"network_ovn_external_routes_remove",
			"tpm_device_type",
			"storage_zfs_clone_copy_rebase",
			"gpu_mdev",
			"resources_pci_iommu",
			"resources_network_usb",
			"resources_disk_address",
			"network_physical_ovn_ingress_mode",
			"network_ovn_dhcp",
			"network_physical_routes_anycast",
			"projects_limits_instances",
			"network_state_vlan",
			"instance_nic_bridged_port_isolation",
			"instance_bulk_state_change",
			"network_gvrp",
			"instance_pool_move",
			"gpu_sriov",
			"pci_device_type",
			"storage_volume_state",
			"network_acl",
			"migration_stateful",
			"disk_state_quota",
			"storage_ceph_features",
			"projects_compression",
			"projects_images_remote_cache_expiry",
			"certificate_project",
			"network_ovn_acl",
			"projects_images_auto_update",
			"projects_restricted_cluster_target",
			"images_default_architecture",
			"network_ovn_acl_defaults",
			"gpu_mig",
			"project_usage",
			"network_bridge_acl",
			"warnings",
			"projects_restricted_backups_and_snapshots",
			"clustering_join_token",
			"clustering_description",
			"server_trusted_proxy",
			"clustering_update_cert",
			"storage_api_project",
			"server_instance_driver_operational",
			"server_supported_storage_drivers",
			"event_lifecycle_requestor_address",
			"resources_gpu_usb",
			"clustering_evacuation",
			"network_ovn_nat_address",
			"network_bgp",
			"network_forward",
			"custom_volume_refresh",
			"network_counters_errors_dropped"
		],
		"api_status": "stable",
		"api_version": "1.0",
		"auth": "trusted",
		"public": false,
		"auth_methods": [
			"tls"
		],
		"environment": {
			"addresses": [],
			"architectures": [
				"x86_64",
				"i686"
			],
			"certificate": "-----BEGIN CERTIFICATE-----\nMIICDDCCAZOgAwIBAgIRANM7xN3/CSoZymlnsdZBV9AwCgYIKoZIzj0EAwMwNzEc\nMBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEXMBUGA1UEAwwOcm9vdEBuaXhv\nLXQxNHMwHhcNMjEwODE1MTczMzIxWhcNMzEwODEzMTczMzIxWjA3MRwwGgYDVQQK\nExNsaW51eGNvbnRhaW5lcnMub3JnMRcwFQYDVQQDDA5yb290QG5peG8tdDE0czB2\nMBAGByqGSM49AgEGBSuBBAAiA2IABAsNPVjM2jdLidIEVlDyLuuzZhhYHETb5aNd\n8u1XEaZorCyTHQERs7lwaqyVM5dmCqi5u/wQQY7oCkEdPVP3swsWtrlQGUGexbhc\nVq5fb/pZ7mrpWdtincCLhTPNZPB1DaNjMGEwDgYDVR0PAQH/BAQDAgWgMBMGA1Ud\nJQQMMAoGCCsGAQUFBwMBMAwGA1UdEwEB/wQCMAAwLAYDVR0RBCUwI4IJbml4by10\nMTRzhwR/AAABhxAAAAAAAAAAAAAAAAAAAAABMAoGCCqGSM49BAMDA2cAMGQCMGrY\nKugCDHfjqEqMtp9mICxXWPLtOkgdbI0gmzorhxeuqghrEK2dulZJS+pQXnNYmwIw\ndlM2Nc7+4VXsD5ZzRhn8YbKdLYE9VgJ2K313nWOl0emagu+ZGYhoYKfpmS7Py6Po\n-----END CERTIFICATE-----\n",
			"certificate_fingerprint": "ab82c644fd3a4292ca56ecf4a277ab1c30551dd7935d32a1c661963bf77a430e",
			"driver": "lxc",
			"driver_version": "4.0.10",
			"firewall": "xtables",
			"kernel": "Linux",
			"kernel_architecture": "x86_64",
			"kernel_features": {
				"netnsid_getifaddrs": "true",
				"seccomp_listener": "true",
				"seccomp_listener_continue": "true",
				"shiftfs": "false",
				"uevent_injection": "true",
				"unpriv_fscaps": "true"
			},
			"kernel_version": "5.14.8",
			"lxc_features": {
				"cgroup2": "true",
				"devpts_fd": "true",
				"idmapped_mounts_v2": "true",
				"mount_injection_file": "true",
				"network_gateway_device_route": "true",
				"network_ipvlan": "true",
				"network_l2proxy": "true",
				"network_phys_macvlan_mtu": "true",
				"network_veth_router": "true",
				"pidfd": "true",
				"seccomp_allow_deny_syntax": "true",
				"seccomp_notify": "true",
				"seccomp_proxy_send_notify_fd": "true"
			},
			"os_name": "NixOS",
			"os_version": "21.11",
			"project": "default",
			"server": "lxd",
			"server_clustered": false,
			"server_name": "nixos-t14s",
			"server_pid": 2238,
			"server_version": "4.18",
			"storage": "dir",
			"storage_version": "1",
			"storage_supported_drivers": [
				{
					"Name": "btrfs",
					"Version": "5.14.1",
					"Remote": false
				},
				{
					"Name": "dir",
					"Version": "1",
					"Remote": false
				}
			]
		}
	}
Creating ubuntu-vm
DBUG[10-03|21:15:56] Connecting to a remote simplestreams server URL=https://images.linuxcontainers.org
DBUG[10-03|21:15:56] Connected to the websocket: ws://unix.socket/1.0/events
DBUG[10-03|21:15:56] Sending request to LXD                   method=POST url=http://unix.socket/1.0/instances etag=
DBUG[10-03|21:15:56]
	{
		"architecture": "",
		"config": {},
		"devices": {},
		"ephemeral": false,
		"profiles": null,
		"stateful": false,
		"description": "",
		"name": "ubuntu-vm",
		"source": {
			"type": "image",
			"certificate": "",
			"alias": "ubuntu/hirsute",
			"server": "https://images.linuxcontainers.org",
			"protocol": "simplestreams",
			"mode": "pull"
		},
		"instance_type": "",
		"type": "virtual-machine"
	}
DBUG[10-03|21:15:57] Got operation from LXD
DBUG[10-03|21:15:57]
	{
		"id": "46f92457-ad46-401e-917b-b89fc68a9c77",
		"class": "task",
		"description": "Creating instance",
		"created_at": "2021-10-03T21:15:57.056964456-06:00",
		"updated_at": "2021-10-03T21:15:57.056964456-06:00",
		"status": "Running",
		"status_code": 103,
		"resources": {
			"instances": [
				"/1.0/instances/ubuntu-vm"
			]
		},
		"metadata": null,
		"may_cancel": false,
		"err": "",
		"location": "none"
	}
DBUG[10-03|21:15:57] Sending request to LXD                   method=GET url=http://unix.socket/1.0/operations/46f92457-ad46-401e-917b-b89fc68a9c77 etag=
DBUG[10-03|21:15:57] Got response struct from LXD
DBUG[10-03|21:15:57]
	{
		"id": "46f92457-ad46-401e-917b-b89fc68a9c77",
		"class": "task",
		"description": "Creating instance",
		"created_at": "2021-10-03T21:15:57.056964456-06:00",
		"updated_at": "2021-10-03T21:15:57.056964456-06:00",
		"status": "Running",
		"status_code": 103,
		"resources": {
			"instances": [
				"/1.0/instances/ubuntu-vm"
			]
		},
		"metadata": null,
		"may_cancel": false,
		"err": "",
		"location": "none"
	}
Error: Failed instance creation: Failed creating instance record: Instance type "virtual-machine" is not supported on this server
```

### Notify maintainers
@fpletz @wucke13 @marsam

### Metadata
```console
 ✘ asbachb@nixos-t14s  /etc/nixos/modules  nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.14.8, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.15`
 - channels(asbachb): `"nixos-21.11pre304061.dd14e5d78e9, nixos-20.09-20.09.4393.068984c00e0"`
 - channels(root): `"nixos-21.11pre319562.c21ba4f7bb4, nixos-hardware, nixos-unstable"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
* lxd
# a list of nixos modules affected by the problem
module:
* lxd
```

### PR DESCRIPTION
**Pull request by adamcstephens**
###### Description of changes

This has been requested over the years, has been attempted in #105651 and #166858, and updated yet again here. Some capability already exists in the merged, linked, PRs.

* Rename existing LXD images in release scripts
* Extract metadata and common configuration from lxc-container module for reuse with lxd-virtual-machine
* Add new lxd-virtual-machine module
* Add new virtual-machine release script
* Update lxd module to add required vhost_vsock kernel module
* Eliminate `with lib` in lxc-container
* Add note to release notes :)

I'm not sure if aarch64-linux workers support nested virtual machines, but I have no platform that has this capability myself either.

Open to any and all feedback. Let's get this merged and finally bring our LXD support up to speed! Just in time for the fork to [Incus](https://linuxcontainers.org/incus/introduction/). :)

Resolves #114194
Resolves #140431

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by adamcstephens** [comment]
@ofborg test lxd

**Comment by adamcstephens** [comment]
Ready for review. I appreciate your time @marsam and/or @AndersonTorres

**Review by AndersonTorres — APPROVED** [review]
LGTM

**Comment by adamcstephens** [comment]
@marsam Any thoughts on this? I'd love to get this across the finish line.

File: nixos/maintainers/scripts/lxd/lxd-virtual-machine-image-inner.nix [hunk 1]
```diff
@@ -0,0 +1,95 @@
+# Edit this configuration file to define what should be installed on
+# your system.  Help is available in the configuration.nix(5) man page
+# and in the NixOS manual (accessible by running ‘nixos-help’).
```
Comment (marsam) [inline nixos/maintainers/scripts/lxd/lxd-virtual-machine-image-inner.nix:3]: Would you mind removing the comments below. I believe it should be enough with the reference to `configuration.nix(5)`

**Review by marsam — APPROVED** [review]
LGTM. Great work, Thanks a lot!

( Sorry for the delay, I messed up my notifications )

File: nixos/maintainers/scripts/lxd/lxd-virtual-machine-image-inner.nix (hunk 1 above)
Comment (adamcstephens) [reply nixos/maintainers/scripts/lxd/lxd-virtual-machine-image-inner.nix:3]: Yep, I've stripped the bulk of the comments from the two image inner files.

**Pull request merged**